### PR TITLE
[SPARK-40594][SQL] Eagerly release hashed relation in ShuffledHashJoin

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

ShuffledHashJoin releases the built hashed relation at the end of task using taskCompletionListener. It is not always good enough for complex sql query.

If a smj or window on the top of the shj without a shuffle, then the hashed relation in shj would be leak. All rows have been consumed in sort before smj or window then the buffer can not allocate the memory which is hold by hashed relation. Then it causes unnecessary spill.

It is a common case in multi-join, since AQE supports convert smj to shj at runtime.

This pr cleanup resource for two cases:
1. after building hashed relation, cleanup build plan
2. after inserting rows in sort, cleanup child

### Why are the changes needed?

Eagerly release hashed relation to avoid unnecessary spill in other memory consumers.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

A local example to see the difference of spill size:

```scala
./bin/spark-shell --driver-memory 512M --conf spark.sql.adaptive.enabled=false --conf spark.sql.shuffle.partitions=10

val fake = 1.to(50).mkString("")
val df1 = spark.range(500).selectExpr("id % 17 as c1")
val df2 = spark.range(100000).selectExpr("id % 17 as c2", s"'$fake' as c3")
val df = df1.join(df2.hint("shuffle_hash"), $"c1" === $"c2").selectExpr("rank() over (partition by c1 order by c2) as r1", "c3")
df.write.format("noop").mode("append").save
```

before:
image

after:
image

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @cloud-fan @JoshRosen @c21  thank you

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala [hunk 1]
```diff
@@ -542,6 +542,8 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
    * Cleans up the resources used by the physical operator (if any). In general, all the resources
    * should be cleaned up when the task finishes but operators like SortMergeJoinExec and LimitExec
    * may want eager cleanup to free up tight resources (e.g., memory).
+   *
+   * Note that, this method is called at executor side.
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala:546]: If you don't mind, could you revert this change to keep this PR simpler?

**Comment by dongjoon-hyun** [comment]
BTW, this is an improvement, `leak` sounds like a bug. Did you want to say `spilled` instead of `leak`?
> then the hashed relation in shj would be leak.

**Comment by ulysses-you** [comment]
@dongjoon-hyun I do not say it is a bug because in general it would not cause task failure just unnecessary spill. It acutally causes memory leak before the task completion. At that time, other memory consumer can not get the memory which they could get.

**Comment by dongjoon-hyun** [comment]
@ulysses-you . Are you sure that this is a definition of memory leak?
> It acutally causes memory leak before the task completion. At that time, other memory consumer can not get the memory which they could get.

**Comment by ulysses-you** [comment]
@dongjoon-hyun ah, I see your concern. I'm not saying it is the definition but is the result. If you think the word `leak` is overkill, I'm fine to revise it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala [hunk 2]
```diff
@@ -88,6 +90,9 @@ case class ShuffledHashJoinExec(
       ignoresDuplicatedKey = ignoreDuplicatedKey)
     buildTime += NANOSECONDS.toMillis(System.nanoTime() - start)
     buildDataSize += relation.estimatedSize
+    hashRelation = relation
+    // cleanup build plan resource eagerly since all rows have been consumed
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala:94]: is it possible to release the memory by hash join itself instead of asking several downstream operators to do so? e.g. release the hash relation if the stream side iterator has been consumed up.

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala:94]: I thought about it. It needs more code change if we want to do this in shj itself.

shj does not have a unified scanner compared with smj, so if we want to release hash relation at the end of output itrator, we need to wrap a new Iterator for shj. Then change all join types with codegen non-codegen code path to use the new Iterator. It will be a code refactor.
